src/gridcoin/mrc.cpp

```diff
@@ -0,0 +1,382 @@
+#include "gridcoin/claim.h"
+#include "gridcoin/tally.h"
+#include "gridcoin/beacon.h"
+#include "gridcoin/quorum.h"
+#include "gridcoin/researcher.h"
+#include "util.h"
+#include "wallet/wallet.h"
+
+using namespace GRC;
+
+namespace {
+//!
+//! \brief Get the hash of a subset of the data in the mrc object used as
+//! input to sign or verify a research reward claim.
+//!
+//! \param mrc           MRC to generate a hash for.
+//! \param last_block_hash Hash of the block at the head of the chain.
+//! \param mrc    mrc transaction of the block that contains
+//! the claim.
+//!
+//! \return Hash of the CPID and last block hash contained in the claim.
+//!
+uint256 GetMRCHash(
+    const MRC& mrc,
+    const uint256& last_block_hash)
+{
+    const CpidOption cpid = mrc.m_mining_id.TryCpid();
+
+    if (!cpid) {
+        return uint256();
+    }
+
+        CHashWriter hasher(SER_NETWORK, PROTOCOL_VERSION);
+
+        hasher << *cpid << last_block_hash;
+
+        return hasher.GetHash();
+}
+} // anonymous namespace
+
+// -----------------------------------------------------------------------------
+// Class: MRC
+// -----------------------------------------------------------------------------
+
+MRC::MRC() : MRC(CURRENT_VERSION)
+{
+}
+
+MRC::MRC(uint32_t version)
+    : m_version(version)
+    , m_research_subsidy(0)
+    , m_magnitude(0)
+    , m_magnitude_unit(0)
+{
+}
+
+bool MRC::WellFormed() const
+{
+    if (m_version <= 0 || m_version > MRC::CURRENT_VERSION) {
+        return false;
+    }
+
+    if (m_version == 1) {
+        return true;
+    }
+
+    if (!m_mining_id.Valid()) {
+        return false;
+    }
+
+    if (m_client_version.empty()) {
+        return false;
+    }
+
+    if (m_mining_id.Which() == MiningId::Kind::CPID) {
+        if (m_research_subsidy <= 0 || m_signature.empty()) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool MRC::HasResearchReward() const
+{
+    return m_mining_id.Which() == MiningId::Kind::CPID;
+}
+
+CAmount MRC::ComputeMRCFee() const
+{
+    CAmount fee = 0;
+
+    // This is the 14 days where fees will be 100% if someone tries to slip an MRC through.
+    const int64_t& zero_payout_interval = Params().GetConsensus().MRCZeroPaymentInterval;
+
+    // Initial fee fraction at end of zero_payout_interval (beginning of valid MRC interval). This is expressed as
+    // separate numerator and denominator for integer math. This is equivalent to 40% fees at the end of the
+    // zero_payout_interval
+    const Fraction& fee_fraction = Params().GetConsensus().InitialMRCFeeFractionPostZeroInterval;
+
+    const CpidOption cpid = m_mining_id.TryCpid();
+
+    if (!cpid) return fee;
+
+    const ResearchAccount& account = Tally::GetAccount(*cpid);
+    const int64_t last_reward_time = account.LastRewardTime();
+
+    // Get the block index of the head of the chain at the time the MRC was filled out.
+    CBlockIndex* prev_block_pindex = mapBlockIndex[m_last_block_hash];
```

Comment:
Consider using `find` and then dereference to avoid inserting data into the map on lookup failures.

Reply:
Done.